pkg/kv/kvserver/client_mvcc_gc_test.go

```diff
@@ -68,3 +89,198 @@ func TestMVCCGCCorrectStats(t *testing.T) {
 	require.Equal(t, oldKeyBytes, newStats.KeyBytes)
 	require.Equal(t, oldValBytes, newStats.ValBytes)
 }
+
+// TestSystemSpanConfigProtectionPoliciesApplyAfterGC is a regression test for
+// https://github.com/cockroachdb/cockroach/issues/113867. This test attempts
+// to recreate the following observed timeline:
+// - Drop table @ t=1
+// - SCHEMA CHANGE GC Job writes range deletion @ t=1, waits for span to be empty
+// - Write a cluster-wide PTS @ t=2
+// - MVCC GC Runs up to t=2
+// - Table span is now empty, SCHEMA CHANGE GC JOB sees span is empty, and
+// deletes span configuration
+// - Range with empty span configuration, falls back to the fallback span
+// configuration and does not respect the cluster-wide protection thereby
+// allowing its GC threshold past t2
+// - Read @ t2 fails with BatchTimestampBeforeGCError
+//
+// The last two steps are where the bug lies as the range should have respected
+// the cluster-wide protection and not allowed its GC threshold to be set past
+// t2.
+func TestSystemSpanConfigProtectionPoliciesApplyAfterGC(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	defer gcjob.SetSmallMaxGCIntervalForTest()()
+
+	ctx := context.Background()
+	var args base.TestClusterArgs
+	args.ServerArgs.DisableDefaultTestTenant = true
+	args.ServerArgs.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
+	args.ServerArgs.Knobs.SpanConfig = &spanconfig.TestingKnobs{
+		OverrideFallbackConf: func(config roachpb.SpanConfig) roachpb.SpanConfig {
+			overrideCfg := config
+			overrideCfg.GCPolicy.TTLSeconds = 1
+			return overrideCfg
+		},
+	}
+	tc := testcluster.StartTestCluster(t, 1, args)
+	defer tc.Stopper().Stop(ctx)
+	sqlDB := tc.ServerConn(0)
+
+	st := tc.Server(0).ClusterSettings()
+	gcjob.EmptySpanPollInterval.Override(ctx, &st.SV, 100*time.Millisecond)
+
+	// Speed up propagation of span configs and protected timestamps.
+	_, err := sqlDB.Exec("SET CLUSTER SETTING kv.closed_timestamp.target_duration = '100ms'")
+	require.NoError(t, err)
+	_, err = sqlDB.Exec("SET CLUSTER SETTING kv.protectedts.poll_interval = '10ms';")
+	require.NoError(t, err)
+
+	// Infrastructure to create a protected timestamp record.
+	execCfg := tc.Server(0).ExecutorConfig().(sql.ExecutorConfig)
+	ptp := execCfg.ProtectedTimestampProvider
+	insqlDB := execCfg.InternalDB
```

Comment:
For master, I would probably have mkRecordAndProtect take an ApplicationLayerInterface and then put these inside that function.